Support schematic data (#73)
ably/kafka-connect-ably: README.md

    data: {"id":"1543960661:0:2","clientId":"kafka-connect-ably-example","connectionId":"SuJTceISnT","timestamp":1623496744538,"encoding":"base64", "channel":"kafka-connect-ably-example","data":"bWVzc2FnZSAz","name":"sink"}
    ```

#### Publishing messages with schema

> Ably Kafka connector supports messages with schema by converting schematic messages to JSON. Produced JSON is then passed as data to Ably message. JSON is produced by eliminating schema fields from the original message.

Reviewer: > JSON is produced by eliminating schema fields from the original message.

Can you clarify what you mean by "eliminate"?

Author: I meant to convey the message to 'get rid of type and other informations from schema' but when I look, I think it's not clear. I made a change here a4942213ba9bd4e516e3ea8afb110f323453a9f4 to hopefully make it a bit more clear but honestly I couldn't find a way to explain this better. I would be open to any suggestions.

For example if you use `AvroConverter` and register an Avro schema to your schema registry, you can use the following command providing your own schema and data:

```shell
 docker-compose exec -T schema-registry kafka-avro-console-producer \

>    --topic topic1 \

Author: Added a slightly lengthy description  d334853024aedb3296fbc5395731852f71b1a43b

   --broker-list kafka:9092 \
   --property key.schema='{"type":"string"}' \
   --property parse.key=true \
   --property key.separator=":" \
   --property value.schema='{"type":"record","name":"myrecord","fields":[{"name":"count","type":"int"}]}' \
   --property schema.registry.url=http://schema-registry:8081 <<EOF
"key1":{"count":1}
"key2":{"count":2}
"key3":{"count":3}
EOF
```

You should receive following JSON messages where you subscribed:

```json

> {"count":1}

Reviewer: The user will also get the Avro-formatted key base64-encoded in the extras, for example:

```json
{
  "id": "1543960661:0:5",
  "clientId": "kafka-connect-ably-example",
  "connectionId": "sY0XR-p0oW",
  "timestamp": 1653912887495,
  "encoding": "json",
  "extras": {
    "kafka": {
      "key": "AAAAAAEIa2V5Mw=="
    }
  },
  "data": "{\"count\":3}"
}
```

I think it's useful to demonstrate that here.

Author: Added an example that contains whole message d5edbd101b440c36017a5ebf16ebcc2f95cbe24c

    data: {"id":"1543960661:0:2","clientId":"kafka-connect-ably-example","connectionId":"SuJTceISnT","timestamp":1623496744538,"encoding":"base64", "channel":"kafka-connect-ably-example","data":"bWVzc2FnZSAz","name":"sink"}
    ```

#### Publishing messages with schema

The Ably Kafka Connector supports messages which contain schema information by converting them to JSON before publishing them to Ably.

> For example, if messages on the Kafka topic are serialized using Avro with schemas registered at https://<your-schema-registry-host>, then set the following configuration so that those messages are converted from Avro to JSON:

Reviewer 2: While I think this Avro example is useful, we should remember that there's nothing Avro-specific about this connector. We support Kafka Connect schemata, which can be generated using either the Avro, JSON or Protobuf converters. We should probably link to the [Kafka Connect Schema Registry docs](https://docs.confluent.io/platform/current/schema-registry/connect.html) so that users can find the appropriate config for their setup?

value.converter=io.confluent.connect.avro.AvroConverter
value.converter.schema.registry.url=https://<your-schema-registry-host>

> `

Reviewer: This should be surrounded with three backticks ("```") to make it a code block

Author: 63ac487f62b51c8b32d382202735ab9e4ed48039

Reviewer 3: Thanks. I'm resolving this conversation on @lmars' behalf as I can see that @ikbalkaya has fixed this.

    data: {"id":"1543960661:0:2","clientId":"kafka-connect-ably-example","connectionId":"SuJTceISnT","timestamp":1623496744538,"encoding":"base64", "channel":"kafka-connect-ably-example","data":"bWVzc2FnZSAz","name":"sink"}
    ```

#### Publishing messages with schema

> Ably Kafka connector supports messages with schema by converting schematic messages to JSON. Produced JSON is then passed as data to Ably message. JSON is produced by mapping schema field names to their values.

Reviewer: ```suggestion
The Ably Kafka Connector supports messages which contain schema information by converting them to JSON before publishing them to Ably.
```

> For example if you use [Kafka Connect Avro Converter](https://www.confluent.io/hub/confluentinc/kafka-connect-avro-converter) and register an Avro schema to your [schema registry](https://docs.confluent.io/platform/current/schema-registry/connect.html), you can use the following. 

Reviewer: Suggested change:

---
For example, if messages on the Kafka topic are serialized using Avro with schemas registered at `https://<your-schema-registry-host>`, then set the following configuration so that those messages are converted from Avro to JSON:

```
value.converter=io.confluent.connect.avro.AvroConverter
value.converter.schema.registry.url=https://<your-schema-registry-host>
```

Author: 513e2e39a2373faf0be4792da8492b32bd68cd9d

> Based on [docker-compose](/docker-compose.yml),  produce a set of test messages specifying `value.schema` property to a valid [Avro Schema](https://avro.apache.org/docs/current/spec.html) and specifying a valid`schema.registry.url` using the `kafka-avro-console-producer` CLI tool. Note that `kafka-avro-console-producer` tool is a part of schema-registry. Replace `<kafka-topic-name>` with one of the topics set in your configuration file.

Reviewer: ```suggestion
If you're running the Ably Kafka Connector locally using Docker Compose as outlined above, then you can use the [`kafka-avro-console-producer` CLI](https://docs.confluent.io/platform/current/tutorials/examples/clients/docs/kafka-commands.html#produce-avro-records) to test producing Avro serialized messages by running the following:
```

```shell
 docker-compose exec -T schema-registry kafka-avro-console-producer \
   --topic topic1 \
   --broker-list kafka:9092 \
   --property key.schema='{"type":"string"}' \
   --property parse.key=true \
   --property key.separator=":" \
   --property value.schema='{"type":"record","name":"myrecord","fields":[{"name":"count","type":"int"}]}' \
   --property schema.registry.url=http://schema-registry:8081 <<EOF
"key1":{"count":1}
"key2":{"count":2}
"key3":{"count":3}
EOF
```

You should receive following JSON messages where you subscribed:

```json
{"count":1}
```

```json
{"count":2}
```

```json
{"count":3}
```

> You will also receive an Avro-formatted key base64 encoded in the extras. For example:

Reviewer: I think we should just have one example that shows the data and the keys in extras coming through, rather than one example with just data, and a second with the keys.

```json
{
	"clientId": "Ably-Kafka-Connector",
	"connectionId": "VSuDXysgaz",
	"data": {
		"count": 1
	},
	"extras": {

> 		"jsonObject": {

Reviewer: This `jsonObject` field should not be here.

Author: @lmars On subscriber side, this JSOn is produced when I try to JSONify Java Object to JSON and looks like there is a `jsonObject` under extras, which then includes the jSON, From your your comment I assume that this area should be skipped when serializing. Can you please confirm that?
<img width="835" alt="Screenshot 2022-05-31 at 07 37 43" src="https://user-images.githubusercontent.com/421091/171108792-c5db2b51-da38-4ba9-9c15-e92b9b20870d.png">

Author: Removed `jsonObject` node e927f3c0a7163aace284830fc7be75eeabf7b9db but I will wait for the response for above to conclude.

Reviewer: > this JSOn is produced when I try to JSONify Java Object to JSON

Ok, I'm not sure what Java is doing there, but `jsonObject` doesn't appear in the underlying message coming out of Ably.

Easiest way to see raw messages is to subscribe via SSE as shown in the other example, and you'll see there won't be any `jsonObject` in there.
